tests/themis/wrappers/android/com/cossacklabs/themis/test/SecureCellSealWithPassphraseTest.java

```diff
@@ -0,0 +1,363 @@
+        SecureCell.Seal cellMK = SecureCell.SealWithKey(secret);
+        SecureCell.Seal cellPW = SecureCell.SealWithPassphrase(secret.toByteArray());
+        byte[] message = "All your base are belong to us!".getBytes(StandardCharsets.UTF_8);
+
+        byte[] encryptedMK = cellMK.encrypt(message);
+        byte[] encryptedPW = cellPW.encrypt(message);
+
+        try {
+            cellPW.decrypt(encryptedMK);
+            fail("expected SecureCellException");
+        }
+        catch (SecureCellException ignored) {}
+        try {
+            cellMK.decrypt(encryptedPW);
+            fail("expected SecureCellException");
+        }
+        catch (SecureCellException ignored) {}
+    }
+
+    @Test
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    public void passphrasesNotCompatibleWithOldAPI() throws SecureCellException {
+        // Old 'passphrase-like' API is not passphrase API at all. Don't use it.
+        SecureCell cellOld = new SecureCell("day 56 of the Q", SecureCell.MODE_SEAL);
+        SecureCell.Seal cellNew = SecureCell.SealWithPassphrase("day 56 of the Q", StandardCharsets.UTF_16);
+        byte[] message = "All your base are belong to us!".getBytes(StandardCharsets.UTF_8);
+
+        byte[] encryptedOld = cellOld.protect("", message).getProtectedData();
+        byte[] encryptedNew = cellNew.encrypt(message);
+
+        try {
+            cellNew.decrypt(encryptedOld);
+            fail("expected SecureCellException");
+        }
+        catch (SecureCellException ignored) {}
+        try {
+            cellOld.unprotect("", new SecureCellData(encryptedNew, null));
+            fail("expected SecureCellException");
+        }
+        catch (SecureCellException ignored) {}
+    }
+
+    @Test
+    public void oldPassphraseAPIIsActuallyUTF16Key() throws SecureCellException {
+        // Yes, it's so weird due to hysterical raisins. So don't use it, really.
+        SecureCell cellOld = new SecureCell("day 56 of the Q", SecureCell.MODE_SEAL);
+        SecureCell.Seal cellNew = SecureCell.SealWithKey("day 56 of the Q".getBytes(StandardCharsets.UTF_16));
+        byte[] message = "All your base are belong to us!".getBytes(StandardCharsets.UTF_8);
+
+        byte[] encryptedOld = cellOld.protect("", message).getProtectedData();
+        byte[] encryptedNew = cellNew.encrypt(message);
+
+        byte[] decryptedOld = cellNew.decrypt(encryptedOld);
+        byte[] decryptedNew = cellOld.unprotect("", new SecureCellData(encryptedNew, null));
+
+        assertArrayEquals(message, decryptedOld);
+        assertArrayEquals(message, decryptedNew);
+    }
+
+    @Test
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    public void passphraseEncodingFailure() {
+        try {
+            // It is an error if the passphrase cannot be represented in the requested charset
+            // without data loss.
+            SecureCell.SealWithPassphrase("пароль", StandardCharsets.US_ASCII);
```

Comment:
so here we test that [this method](https://github.com/cossacklabs/themis/pull/635/files#diff-7a91dbdd4b6a8a3c6b48e393d7ca99cfR39) will fail if receive inconsistent string and encoding?

Reply:
> so here we test that this method will fail if receive inconsistent string and encoding?

Yeah, that's the idea behind this test.

`PassphraseBytes#encodePassphrase` is not directly tested, but here we ensure that it's actually used, not `String#getBytes`. Encoding conversion happy path is indirectly verified by other passphrase tests.